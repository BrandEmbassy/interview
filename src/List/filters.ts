import { sortBy, prop, reverse } from 'ramda';
import { Contact } from '../state/types';

export enum FILTER {
  All,
  Ascending,
  Descending,
}

export type FilterSort = (list: Contact[]) => Contact[];
export interface FilterType {
  name: string;
  filter: FILTER;
  sort: FilterSort;
}
export const Filters: FilterType[] = [
  {
    name: 'All',
    filter: FILTER.All,
    sort: (list: Contact[]): Contact[] => list,
  },
  {
    name: 'A-Z',
    filter: FILTER.Ascending,
    sort: (list: Contact[]): Contact[] => sortBy(prop('name'))(list),
  },
  {
    name: 'Z-A',
    filter: FILTER.Descending,
    sort: (list: Contact[]): Contact[] => reverse(sortBy(prop('name'))(list)),
  },
];
