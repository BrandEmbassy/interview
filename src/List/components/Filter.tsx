import * as React from 'react';
import * as cx from 'classnames';
import { FILTER, Filters, FilterType } from '../filters';
import { switchFilter } from '../state';

interface Props {
  activeFilter: FILTER;
}

const Filter: React.SFC<Props> = ({ activeFilter }) => {
  return (
    <div className="filter">
      {Filters.map((filter: FilterType) => (
        <div
          key={`filter-${filter.filter}`}
          className={cx('filter__item', {
            'filter__item--active': activeFilter === filter.filter,
          })}
          onClick={() => switchFilter(filter.filter)}
        >
          {filter.name}
        </div>
      ))}
    </div>
  );
};

export default Filter;
