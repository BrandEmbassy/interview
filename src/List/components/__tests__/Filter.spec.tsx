import * as React from 'react';
import Filter from '../Filter';
import render from '../../../test/render';
import { FILTER } from '../../filters';

Object.keys(FILTER)
  .map(fKey => ({ activeFilter: FILTER[fKey] }))
  .forEach(props => {
    it(`should render Filter - ${JSON.stringify(props, null, 2)}`, () => {
      expect(render(<Filter {...props} />)).toMatchSnapshot();
    });
  });
