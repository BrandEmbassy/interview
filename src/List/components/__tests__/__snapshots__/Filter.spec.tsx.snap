// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Filter - {
  "activeFilter": "All"
} 1`] = `
<div
  className="filter"
>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    All
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    A-Z
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    Z-A
  </div>
</div>
`;

exports[`should render Filter - {
  "activeFilter": "Ascending"
} 1`] = `
<div
  className="filter"
>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    All
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    A-Z
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    Z-A
  </div>
</div>
`;

exports[`should render Filter - {
  "activeFilter": "Descending"
} 1`] = `
<div
  className="filter"
>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    All
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    A-Z
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    Z-A
  </div>
</div>
`;

exports[`should render Filter - {
  "activeFilter": 0
} 1`] = `
<div
  className="filter"
>
  <div
    className="filter__item filter__item--active"
    onClick={[Function]}
  >
    All
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    A-Z
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    Z-A
  </div>
</div>
`;

exports[`should render Filter - {
  "activeFilter": 1
} 1`] = `
<div
  className="filter"
>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    All
  </div>
  <div
    className="filter__item filter__item--active"
    onClick={[Function]}
  >
    A-Z
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    Z-A
  </div>
</div>
`;

exports[`should render Filter - {
  "activeFilter": 2
} 1`] = `
<div
  className="filter"
>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    All
  </div>
  <div
    className="filter__item"
    onClick={[Function]}
  >
    A-Z
  </div>
  <div
    className="filter__item filter__item--active"
    onClick={[Function]}
  >
    Z-A
  </div>
</div>
`;
